src/charm.py

```diff
@@ -0,0 +1,1624 @@
+        """
+        if not self.unit.is_leader():
+            return
+        vault = self._get_active_vault_client()
+        if vault is None:
+            logger.warning("Vault is not active, cannot generate autounseal credentials")
+            return
+
+        vault.enable_secrets_engine(SecretsBackend.TRANSIT, AUTOUNSEAL_MOUNT_PATH)
+
+        key_name, approle_id, secret_id = vault.create_autounseal_credentials(
+            relation.id,
+            AUTOUNSEAL_MOUNT_PATH,
+            AUTOUNSEAL_POLICY_PATH,
+        )
+
+        self._set_autounseal_relation_data(relation, key_name, approle_id, secret_id)
+
+    def _sync_vault_autounseal(self) -> None:
+        """Go through all the vault-autounseal relations and send necessary credentials.
+
+        This looks for any outstanding requests for auto-unseal that may have
+        been missed. If there are any, it generates the credentials and sets
+        them in the relation databag.
+        """
+        if not self.unit.is_leader():
+            logger.debug("Only leader unit can handle a vault-autounseal request")
+            return
+        outstanding_requests = self.vault_autounseal_provides.get_outstanding_requests()
+        for relation in outstanding_requests:
+            self._generate_and_set_autounseal_credentials(relation)
+
+    def _set_autounseal_relation_data(
+        self, relation: Relation, key_name: str, approle_id: str, approle_secret_id: str
+    ) -> None:
+        """Set the required autounseal data in the relation databag.
+
+        Args:
+            relation: Relation for which the auto-unseal data is being set
+            key_name: The vault transit key name used for auto-unseal
+            approle_id: The AppRole ID which has permission to use this key
+            approle_secret_id: The AppRole secret ID
+        """
+        vault_address = self._get_relation_api_address(relation)
+        if not vault_address:
+            logger.warning("Vault address not available, ignoring request to set autounseal data")
+            return
+        ca_cert = (
+            self.tls.pull_tls_file_from_workload(File.CA)
+            if self.tls.ca_certificate_is_saved()
+            else None
+        )
+        if not ca_cert:
+            logger.warning("CA certificate not available, ignoring request to set autounseal data")
+            return
+
+        self.vault_autounseal_provides.set_autounseal_data(
+            relation,
+            vault_address,
+            AUTOUNSEAL_MOUNT_PATH,
+            key_name,
+            approle_id,
+            approle_secret_id,
+            ca_cert,
+        )
+
+    def _on_install(self, event: InstallEvent):
+        """Handle the install charm event."""
+        if not self._container.can_connect():
+            event.defer()
+            return
+        self._delete_vault_data()
+
+    def _on_collect_status(self, event: CollectStatusEvent):  # noqa: C901
```

Comment:
What's C901?

Reply:
function too complex.
I'm not sure how to split this to get rid of that. It runs all checks required and sets the status of the charm. We follow the single exit hook pattern.